core/src/main/java/io/substrait/relation/physical/HashJoin.java

```diff
@@ -0,0 +1,79 @@
+package io.substrait.relation.physical;
+
+import io.substrait.expression.FieldReference;
+import io.substrait.proto.HashJoinRel;
+import io.substrait.relation.AbstractJoin;
+import io.substrait.relation.RelVisitor;
+import io.substrait.type.Type;
+import io.substrait.type.TypeCreator;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Stream;
+import org.immutables.value.Value;
+
+@Value.Immutable
+public abstract class HashJoin extends AbstractJoin {
+
+  public abstract Optional<List<FieldReference>> getLeftKeys();
+
+  public abstract Optional<List<FieldReference>> getRightKeys();
+
+  public abstract JoinType getJoinType();
+
+  @Override
+  protected Type.Struct deriveRecordType() {
+    Stream<Type> leftTypes =
+        switch (getJoinType()) {
+          case RIGHT, OUTER -> getLeft().getRecordType().fields().stream()
```

Comment:
Should we add RIGHT_SEMI and RIGHT_ANTI? (Same with LEFT below)

Reply:
Good catch. no columns from the opposite are added.